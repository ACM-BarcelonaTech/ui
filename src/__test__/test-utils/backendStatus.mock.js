const backendStatus = {
  gem2: {
    executionArn: '"arn:aws:states:eu-east-1:000000000000:execution:biomage-gem2s-test-test9cd23b9138d3d5b6141835c85101495a92cd:fb047b29-a9fb-4e36-a0fc-b792e243test"',
    paramsHash: '67515f56ad338b8c7339593582614ff62e376587', // pragma: allowlist secret
    stateMachineArn: 'arn:aws:states:eu-east-1:000000000000:stateMachine:biomage-gem2s-test-test9cd23b9138d3d5b6141835c85101495atest',
  },
  pipeline: {
    executionArn: 'arn:aws:states:eu-west-1:000000000000:execution:biomage-qc-test-test9cd23b9138d3d5b6141835c85101495a92cd:aef9d56d-a20c-4c10-ac1b-2c37cb87test',
    stateMachineArn: 'arn:aws:states:eu-west-1:000000000000:execution:biomage-qc-test-test9cd23b9138d3d5b6141835c85101495a92cd:aef9d56d-a20c-4c10-ac1b-2c37cb87test',
    completedSteps: [
      'ClassifierFilter',
      'CellSizeDistributionFilter',
      'MitochondrialContentFilter',
      'NumGenesVsNumUmisFilter',
      'DoubletScoresFilter',
      'DataIntegration',
      'ConfigureEmbedding',
    ],
    startDate: '2021-08-16T12:10:32.320Z',
    stopDate: '2021-08-16T12:21:07.847Z',
    status: 'SUCCEEDED',
  },
  worker: {
    ready: true,
    restartCount: 0,
    started: true,
    status: 'Running',
  },
};

export default backendStatus;
