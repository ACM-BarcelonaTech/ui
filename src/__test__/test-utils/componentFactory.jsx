import React from 'react';
import _ from 'lodash';

const createTestComponentFactory = (Component, defaultProps) => (customProps = {}) => {
  const props = _.merge(
    defaultProps,
    customProps,
  );

  // eslint-disable-next-line react/jsx-props-no-spreading
  return <Component {...props} />;
};

export default createTestComponentFactory;
