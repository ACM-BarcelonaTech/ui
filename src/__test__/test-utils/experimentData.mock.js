import fake from '__test__/test-utils/constants';
import mockBackendStatus from '__test__/test-utils/backendStatus.mock';

const mockExperimentData = {
  data: {
    experimentId: fake.EXPERIMENT_ID,
    experimentName: fake.EXPERIMENT_NAME,
    meta: mockBackendStatus,
    projectId: fake.PROJECT_ID,
    sampleIds: [fake.SAMPLE_ID],
  },
};

export default mockExperimentData;
