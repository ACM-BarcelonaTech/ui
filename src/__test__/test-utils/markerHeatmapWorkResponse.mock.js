const workResponse = {
  data: {
    GENE1: {
      rawExpression: {
        expression: [1, 2, 3, 4, 5],
        mean: 3,
        stdev: 1.41421,
      },
      truncatedExpression: {
        expression: [1, 2, 3, 4, 5],
        min: 0,
        max: 5,
      },
      zScore: [0.1, 0.2, 0.3, 0.4, 0.5],
    },
    GENE2: {
      rawExpression: {
        expression: [1, 2, 3, 4, 5],
        mean: 3,
        stdev: 1.41421,
      },
      truncatedExpression: {
        expression: [1, 2, 3, 4, 5],
        min: 0,
        max: 5,
      },
      zScore: [0.1, 0.2, 0.3, 0.4, 0.5],
    },
    GENE3: {
      rawExpression: {
        expression: [1, 2, 3, 4, 5],
        mean: 3,
        stdev: 1.41421,
      },
      truncatedExpression: {
        expression: [1, 2, 3, 4, 5],
        min: 0,
        max: 5,
      },
      zScore: [0.1, 0.2, 0.3, 0.4, 0.5],
    },
  },
  order: ['GENE1', 'GENE2', 'GENE3'],
};

export default workResponse;
