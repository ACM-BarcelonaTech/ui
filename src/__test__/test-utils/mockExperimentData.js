import fake from '__test__/test-utils/constants';
import mockBackendStatus from '__test__/data/backend_status.json';

const generateMockExperimentData = (experimentId) => ({
  data: {
    experimentId,
    experimentName: `Experiment ${experimentId}`,
    meta: mockBackendStatus,
    projectId: fake.PROJECT_ID,
    sampleIds: [fake.SAMPLE_ID],
  },
});

export default generateMockExperimentData;
