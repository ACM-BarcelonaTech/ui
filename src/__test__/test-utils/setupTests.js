import '@testing-library/jest-dom/extend-expect';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import preloadAll from 'jest-next-dynamic';

configure({ adapter: new Adapter() });
jest.mock('localforage');
jest.mock('utils/pushNotificationMessage');

// Add stuff that needs to run once, before all tests
beforeAll(async () => {
  await preloadAll();
});

// Add stuff that needs to run before each test
beforeEach(async () => {
  console.log('HELLO WORLD ++++++++++');
});
