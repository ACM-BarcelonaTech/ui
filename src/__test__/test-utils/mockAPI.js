import _ from 'lodash';

// Defaul platform data
import cellSetsData from '__test__/data/cell_sets.json';
import backendStatusData from '__test__/data/backend_status.json';
import processingConfigData from '__test__/data/processing_config.json';
import mockExperimentData from '__test__/test-utils/experimentData.mock';

const promiseResponse = (
  response,
  options = {},
) => Promise.resolve(new Response(response, options));

const statusResponse = (code, body) => Promise.resolve({
  status: code,
  body: JSON.stringify(body),
});

const workerResponse = (body, error = false) => promiseResponse(JSON.stringify({
  results: [
    {
      body: JSON.stringify(body),
    },
  ],
  response: { error },
}));

const generateDefaultMockAPIResponses = (mockExperimentId) => ({
  [mockExperimentId]: () => promiseResponse(
    JSON.stringify(mockExperimentData),
  ),
  processingConfig: () => promiseResponse(
    JSON.stringify(processingConfigData),
  ),
  '/cellSets': () => promiseResponse(
    JSON.stringify(cellSetsData),
  ),
  '/backendStatus': () => promiseResponse(
    JSON.stringify(backendStatusData),
  ),
});

const mockApi = (apiMapping) => (req) => {
  const path = req.url;

  const key = _.find(
    Object.keys(apiMapping),
    (urlStub) => path.endsWith(urlStub),
  );

  return apiMapping[key](req);
};

export default mockApi;
export {
  generateDefaultMockAPIResponses,
  promiseResponse,
  statusResponse,
  workerResponse,
};
