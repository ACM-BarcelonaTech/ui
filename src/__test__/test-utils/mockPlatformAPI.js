import _ from 'lodash';

// Defaul platform data
import cellSetsData from '__test__/data/cell_sets.json';
import backendStatusData from '__test__/data/backend_status.json';
import processingConfigData from '__test__/data/processing_config.json';
import mockExperimentData from '__test__/test-utils/experimentData.mock';

const promiseResponse = (response) => Promise.resolve(new Response(response));
const promiseStatus = (code, body) => Promise.resolve({
  status: code,
  body: JSON.stringify(body),
});

const generateMockApiConfig = (mockExperimentId, customMap = {}) => {
  const defaultMockApiMapping = {
    [mockExperimentId]: () => promiseResponse(
      JSON.stringify(mockExperimentData),
    ),
    processingConfig: () => promiseResponse(
      JSON.stringify(processingConfigData),
    ),
    '/cellSets': () => promiseResponse(
      JSON.stringify(cellSetsData),
    ),
    '/backendStatus': () => promiseResponse(
      JSON.stringify(backendStatusData),
    ),
  };

  return _.merge(
    defaultMockApiMapping,
    customMap,
  );
};

const mockApi = (experimentId, customMap = {}) => (req) => {
  const path = req.url;

  const apiMapping = generateMockApiConfig(experimentId, customMap);

  const key = _.find(
    Object.keys(apiMapping),
    (urlStub) => path.endsWith(urlStub),
  );

  return apiMapping[key](req);
};

export default mockApi;
export {
  promiseResponse,
  promiseStatus,
};
