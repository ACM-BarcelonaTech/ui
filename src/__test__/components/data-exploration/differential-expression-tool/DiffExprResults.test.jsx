import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import waitForActions from 'redux-mock-store-await-actions';

import { Empty } from 'antd';
import DiffExprResults from 'components/data-exploration/differential-expression-tool/DiffExprResults';
import { fetchWork } from 'utils/work/fetchWork';
import { DIFF_EXPR_LOADING, DIFF_EXPR_LOADED } from 'redux/actionTypes/differentialExpression';
import '__test__/test-utils/setupTests';

import Loader from 'components/Loader';

jest.mock('utils/getTimeoutForWorkerTask', () => ({
  __esModule: true, // this property makes it work
  default: () => 60,
}));

jest.mock('utils/work/fetchWork', () => ({
  __esModule: true, // this property makes it work
  fetchWork: jest.fn(() => new Promise((resolve) => resolve({
    rows: [
      {
        p_val: 1.4969461240347763e-12, p_val_adj: 1.647289002209057e-11, logFC: -1.4274754343649423, gene_names: 'A', auc: '0.1',
      },
      {
        p_val: 2.4969461240347763e-12, p_val_adj: 2.647289002209057e-11, logFC: -2.4274754343649423, gene_names: 'B', auc: '0.2',
      },
      {
        p_val: 3.4969461240347763e-12, p_val_adj: 3.647289002209057e-11, logFC: -3.4274754343649423, gene_names: 'C', auc: '0.3',
      },
      {
        p_val: 4.4969461240347763e-12, p_val_adj: 4.647289002209057e-11, logFC: -4.4274754343649423, gene_names: 'D', auc: '0.4',
      },
      {
        p_val: 5.4969461240347763e-12, p_val_adj: 5.647289002209057e-11, logFC: -5.4274754343649423, gene_names: 'E', auc: '0.5',
      },
    ],
    total: 500,
  }))),
}));

const mockStore = configureMockStore([thunk]);

const experimentId = '1234';

const backendStatus = {
  [experimentId]: {
    status: {
      pipeline: {
        startDate: '2021-01-01T00:00:00',
      },
    },
  },
};

const resultState = {
  genes: {
    selected: [],
  },
  cellInfo: {
    focus: {
      store: 'genes',
      key: 'C',
    },
  },
  cellSets: {
    loading: false,
    error: false,
    selected: [],
    properties: {
      'cluster-a': {
        name: 'Name of Cluster A',
        key: 'cluster-a',
        cellIds: new Set([1, 2]),
        color: '#00FF00',
      },
      'cluster-b': {
        name: 'Name of Cluster B',
        key: 'cluster-b',
        cellIds: new Set([2, 3, 4, 5]),
        color: '#FF0000',
      },
      'scratchpad-a': {
        cellIds: new Set([3]),
        key: 'scratchpad-a',
        name: 'Name of Scratchpad A',
        color: '#ff00ff',
      },
      louvain: {
        cellIds: new Set(),
        name: 'Louvain clusters',
        key: 'louvain',
        type: 'cellSets',
        rootNode: true,
      },
      scratchpad: {
        cellIds: new Set(),
        name: 'Custom selections',
        key: 'scratchpad',
        type: 'cellSets',
        rootNode: true,
      },
    },
    hierarchy: [
      {
        key: 'louvain',
        children: [{ key: 'cluster-a' }, { key: 'cluster-b' }],
      },
      {
        key: 'scratchpad',
        children: [{ key: 'scratchpad-a' }],
      },
    ],
  },
  differentialExpression: {
    properties: {
      data: [
        {
          p_val: 1.4969461240347763e-12, p_val_adj: 1.647289002209057e-11, logFC: -1.4274754343649423, gene_names: 'A', auc: '0.1',
        },
        {
          p_val: 2.4969461240347763e-12, p_val_adj: 2.647289002209057e-11, logFC: -2.4274754343649423, gene_names: 'B', auc: '0.2',
        },
        {
          p_val: 3.4969461240347763e-12, p_val_adj: 3.647289002209057e-11, logFC: -3.4274754343649423, gene_names: 'C', auc: '0.3',
        },
        {
          p_val: 4.4969461240347763e-12, p_val_adj: 4.647289002209057e-11, logFC: -4.4274754343649423, gene_names: 'D', auc: '0.4',
        },
        {
          p_val: 5.4969461240347763e-12, p_val_adj: 5.647289002209057e-11, logFC: -5.4274754343649423, gene_names: 'E', auc: '0.5',
        },
      ],
      loading: false,
      error: false,
      total: 5,
    },
    comparison: {
      type: 'between',
      group: {
        between: {
          cellSet: 'louvain/cluster-a',
          compareWith: 'louvain/cluster-b',
          basis: 'scratchpad/scratchpad-a',
        },
      },
    },
  },
  backendStatus,
};

const noResultState = {
  ...resultState,
  differentialExpression: {
    ...resultState.differentialExpression,
    properties: {
      ...resultState.differentialExpression.properties,
      data: [],
      loading: false,
      error: false,
      total: 0,
    },
  },
};

const withResultStore = mockStore(resultState);
const noResultStore = mockStore(noResultState);

describe('DiffExprResults', () => {
  it('renders correctly', () => {
    const component = mount(
      <Provider store={withResultStore}>
        <DiffExprResults experimentId={experimentId} onGoBack={jest.fn()} width={100} height={200} />
      </Provider>,
    );

    const table = component.find('Table');
    const spin = component.find(Loader);
    expect(spin.length).toEqual(0);
    expect(table.length).toEqual(1);
    expect(table.getElement().props.columns.length).toEqual(7);
    expect(table.getElement().props.columns[0].key).toEqual('lookup');
    expect(table.getElement().props.columns[1].key).toEqual('gene_names');
    expect(table.getElement().props.columns[2].key).toEqual('logFC');
    expect(table.getElement().props.columns[3].key).toEqual('p_val_adj');
    expect(table.getElement().props.columns[4].key).toEqual('pct_1');
    expect(table.getElement().props.columns[5].key).toEqual('pct_2');
    expect(table.getElement().props.columns[6].key).toEqual('auc');

    expect(table.getElement().props.dataSource.length).toEqual(5);
    expect(table.getElement().props.data.length).toEqual(5);
  });

  it('can sort the gene names in alphabetical order', async () => {
    const newPagination = {
      current: 1,
      pageSize: 4,
      showSizeChanger: true,
      total: 4,
    };

    const newSorter = {
      column: {
        dataIndex: 'gene_names',
        key: 'gene_names',
      },
      render: jest.fn(),
      columnKey: 'gene_names',
      field: 'gene_names',
      order: 'ascend',
    };

    const component = mount(
      <Provider store={withResultStore}>
        <DiffExprResults experimentId={experimentId} onGoBack={jest.fn()} width={100} height={200} />
      </Provider>,
    );

    const table = component.find('Table');

    act(() => {
      table.getElement().props.onChange(newPagination, {}, newSorter);
    });

    // // Wait for side-effect to propagate (properties loading and loaded).
    await waitForActions(withResultStore, [DIFF_EXPR_LOADING, DIFF_EXPR_LOADED]);

    expect(fetchWork).toHaveBeenCalledWith(
      '1234',
      {
        cellSet: 'cluster-a',
        compareWith: 'cluster-b',
        basis: 'scratchpad-a',
        experimentId: '1234',
        name: 'DifferentialExpression',
      },
      withResultStore.getState,
      {
        extras: {
          pagination: {
            limit: 4, offset: 0, orderBy: 'gene_names', orderDirection: 'ASC', responseKey: 0,
          },
        },
        timeout: 60,
      },
    );

    expect(withResultStore.getActions()[0]).toMatchSnapshot();
    expect(withResultStore.getActions()[1]).toMatchSnapshot();
  });

  it('Having a focused gene triggers focused view for `eye` button.', () => {
    // Redefine store from `beforeEach`.
    const component = mount(
      <Provider store={withResultStore}>
        <DiffExprResults experimentId={experimentId} onGoBack={jest.fn()} width={100} height={200} />
      </Provider>,
    );

    const table = component.find('Table');

    table.getElement().props.data.forEach((row) => {
      const lookupComponent = mount(
        <Provider store={withResultStore}>
          {row.lookup}
        </Provider>,
      );

      const focusButtonTooltip = lookupComponent.find('FocusButton Tooltip');

      if (row.gene_names === 'C') {
        expect(focusButtonTooltip.props().title).toContain('Hide');
      } else {
        expect(focusButtonTooltip.props().title).toContain('Show');
      }

      lookupComponent.unmount();
    });
  });

  it('Show comparison settings button works.', () => {
    const component = mount(
      <Provider store={withResultStore}>
        <DiffExprResults experimentId={experimentId} onGoBack={jest.fn()} width={100} height={200} />
      </Provider>,
    );
    const button = component.find('#settingsButton').first();
    expect(button.text()).toContain('Show');
    button.simulate('click');
    expect(button.text()).toContain('Hide');

    const div = component.find('#settingsText');
    // Should display name of cluster instead of ID
    expect(div.text()).toEqual('Name of Cluster A vs. Name of Cluster B in Name of Scratchpad A');
    button.simulate('click');
    expect(button.childAt(0).text()).toEqual('Show settings');
    expect(!div);
  });

  it("Doesn't show loading indicator if there is no data returned", () => {
    const component = mount(
      <Provider store={noResultStore}>
        <DiffExprResults
          experimentId={experimentId}
        />
      </Provider>,
    );

    const spin = component.find(Loader);
    const empty = component.find(Empty);

    // There should be no loader
    expect(spin.length).toEqual(0);

    // Expect table to contain Empty component
    expect(empty.length).toEqual(1);
  });
});
