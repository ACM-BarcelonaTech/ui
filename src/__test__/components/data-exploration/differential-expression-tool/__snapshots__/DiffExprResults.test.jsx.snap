// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiffExprResults can sort the gene names in alphabetical order 1`] = `
Object {
  "payload": Object {
    "experimentId": "1234",
  },
  "type": "diffExpr/loading",
}
`;

exports[`DiffExprResults can sort the gene names in alphabetical order 2`] = `
Object {
  "payload": Object {
    "comparisonGroup": Object {
      "basis": "scratchpad/scratchpad-a",
      "cellSet": "louvain/cluster-a",
      "compareWith": "louvain/cluster-b",
    },
    "comparisonType": "between",
    "data": Array [
      Object {
        "auc": "0.1",
        "gene_names": "A",
        "logFC": -1.4274754343649423,
        "p_val": 1.4969461240347763e-12,
        "p_val_adj": 1.647289002209057e-11,
      },
      Object {
        "auc": "0.2",
        "gene_names": "B",
        "logFC": -2.4274754343649425,
        "p_val": 2.4969461240347763e-12,
        "p_val_adj": 2.647289002209057e-11,
      },
      Object {
        "auc": "0.3",
        "gene_names": "C",
        "logFC": -3.4274754343649425,
        "p_val": 3.4969461240347765e-12,
        "p_val_adj": 3.647289002209057e-11,
      },
      Object {
        "auc": "0.4",
        "gene_names": "D",
        "logFC": -4.4274754343649425,
        "p_val": 4.496946124034777e-12,
        "p_val_adj": 4.647289002209057e-11,
      },
      Object {
        "auc": "0.5",
        "gene_names": "E",
        "logFC": -5.4274754343649425,
        "p_val": 5.4969461240347765e-12,
        "p_val_adj": 5.647289002209057e-11,
      },
    ],
    "experimentId": "1234",
    "total": 500,
  },
  "type": "diffExpr/loaded",
}
`;
